ddtrace/sampler.py

```diff
@@ -348,25 +343,32 @@ def sample(self, span):
         """
         # Go through all rules and grab the first one that matched
         # DEV: This means rules should be ordered by the user from most specific to least specific
+        has_configured_rate_limit = self.limiter.rate_limit != DEFAULT_SAMPLING_RATE_LIMIT
+
+        rule_matched = False
         for rule in self.rules:
             if rule.matches(span):
                 sampler = rule
+                rule_matched = True
                 break
-        else:
-            # No rules match so use agent based sampling
-            return super(DatadogSampler, self).sample(span)
```

Comment:
The removal of this early return is the only part of this diff that seems remotely likely to degrade performance, as far as I can tell.

Reply:
It would make sense why this change would have an impact on the span start/finish micro benchmarks because this code runs on root spans when sampling.

I wonder if it is worth having a start/finish test that doesn't trigger a sampling decision, and/or have a test dedicated to sampling?

It does make me a bit worried about the `scenario:flasksimple-appsec-telemetry` since that one should normalize the expensive cost of sampling / give a more "realistic" scenario... but also why does only that one show an increase?

🤔

Reply:
The latest changes reduce the performance impact to what I think are acceptable levels. What do you think? https://github.com/DataDog/dd-trace-py/pull/6693#issuecomment-1683029594

Reply:
nice, seems like higher memory usage, but that is fine, as long as we don't impact execution time, should be ok.